subsys/sensing/sensor/phy_3d_sensor/phy_3d_sensor.c

```diff
@@ -18,63 +18,410 @@
+	.chan_all = SENSOR_CHAN_GYRO_XYZ,
+	.shift = SENSING_GYRO_Q31_SHIFT,
+	.q31_to_sensor_value = gyro_q31_to_sensor_value,
+	.sensor_value_to_q31 = gyro_sensor_value_to_q31,
+};
+
+static void phy_3d_sensor_data_ready_handler(const struct device *dev,
+		const struct sensor_trigger *trig)
+{
+	struct phy_3d_sensor_context *ctx = CONTAINER_OF(trig,
+			struct phy_3d_sensor_context,
+			trig);
+	int ret;
+
+	LOG_DBG("%s: trigger type:%d", dev->name, trig->type);
+
+	ret = sensor_sample_fetch(ctx->hw_dev);
+	if (ret) {
+		LOG_ERR("%s: sample fetch failed: %d", dev->name, ret);
+		return;
+	}
+
+	sensing_sensor_notify_data_ready(ctx->dev);
+}
+
+static int phy_3d_sensor_enable_data_ready(struct phy_3d_sensor_context *ctx,
+		bool enable)
+{
+	sensor_trigger_handler_t handler;
+	int ret = 0;
+
+	ctx->trig.type = SENSOR_TRIG_DATA_READY;
+	ctx->trig.chan = ctx->custom->chan_all;
+
+	if (enable != ctx->data_ready_enabled) {
+		handler = enable ? phy_3d_sensor_data_ready_handler : NULL;
+
+		ret = sensing_sensor_set_data_ready(ctx->dev, enable);
+		if (!ret) {
+			if (!sensor_trigger_set(ctx->hw_dev, &ctx->trig,
+						handler)) {
+				ctx->data_ready_enabled = enable;
+			} else {
+				sensing_sensor_set_data_ready(ctx->dev,
+						!enable);
+				LOG_INF("%s: set data ready trigger not successfully",
+						ctx->dev->name);
+			}
+		}
+	}
+
+	LOG_DBG("%s: trigger data ready enabled:%d", ctx->dev->name,
+			ctx->data_ready_enabled);
+
+	return ret;
+}
+
 static int phy_3d_sensor_init(const struct device *dev,
 		const struct sensing_sensor_info *info,
 		const sensing_sensor_handle_t *reporter_handles,
 		int reporters_count)
 {
-	return 0;
-}
+	struct phy_3d_sensor_context *ctx;
+	int ret;
 
-static int phy_3d_sensor_deinit(const struct device *dev)
-{
-	return 0;
+	ARG_UNUSED(reporter_handles);
+	ARG_UNUSED(reporters_count);
+
+	ctx = sensing_sensor_get_ctx_data(dev);
+	ctx->dev = dev;
+
+	switch (ctx->sensor_type) {
+	case SENSING_SENSOR_TYPE_MOTION_ACCELEROMETER_3D:
+		ctx->custom = &custom_accel;
+		break;
+	case SENSING_SENSOR_TYPE_MOTION_GYROMETER_3D:
+		ctx->custom = &custom_gyro;
+		break;
+	default:
+		LOG_ERR("phy_3d_sensor doesn't support sensor type %d",
+				ctx->sensor_type);
+		return -ENOTSUP;
+	}
+
+	LOG_INF("%s: Underlying device: %s", dev->name, ctx->hw_dev->name);
+
+	/* Check underlying device supports data ready or not */
+	ret = phy_3d_sensor_enable_data_ready(ctx, true);
+	ctx->data_ready_support = ctx->data_ready_enabled;
+	if (!ret && ctx->data_ready_enabled) {
+		ret = phy_3d_sensor_enable_data_ready(ctx, false);
+	}
+
+	return ret;
 }
 
 static int phy_3d_sensor_read_sample(const struct device *dev,
 		void *buf, int size)
 {
-	return 0;
+	struct phy_3d_sensor_context *ctx;
+	struct sensing_sensor_value_3d_q31 *sample = buf;
+	struct sensor_value value[PHY_3D_SENSOR_CHANNEL_NUM];
+	int i, ret;
+
+	ctx = sensing_sensor_get_ctx_data(dev);
+
+	if (!ctx->data_ready_enabled) {
+		ret = sensor_sample_fetch(ctx->hw_dev);
```

Comment:
This can/will likely cause a lot of false errors. Since the phy_3d_sensor is instantiated for each type and not once for the actual physical sensor. Imagine you have a standard IMU which provides accel/gyro/die_temp. This means you'll have 2 instances for the `phy_3d_sensor` and I'm assuming a `phy_1d_sensor`. If you try to read all 3 separately (in succession), the `fetch` call will fail on the 2nd and 3rd calls since new data isn't available yet. This is why my original design (referencing https://github.com/zephyrproject-rtos/zephyr/issues/58925) calls for only 1 device node and 3 info nodes to be bound in the example above. The problem is still complicated, but at least all the read requests will funnel into the same device instance so it would be easier (since they share the data/config structs).

Reply:
Thanks for your comments. I use `sensor_sample_fetch_chan` instead of `sensor_sample_fetch` now. It can handle this case well.

Reply:
Still won't work because the `fetch` checks the status register, if you don't have a new sample between `fetch` calls, this will fail.

Reply:
Dose `between fetch calls` mean between two `fetch accel` calls or between `fetch accel` call and `fetch gyro` call?

- Case 1: between two `fetch accel` calls
In this case, the sensing subsystem(polling mode) and data ready trigger function(data ready trigger mode) will make sure that there is a new sample between two `fetch accel` calls.

> * Polling mode:
> The freq of `fetch` called by Sensing subsystem is always less than sensor working ODR.
> * Data ready trigger mode:
> The `fetch` is triggerd by the data ready trigger function which set by `sensor_trigger_set`.
> And data ready trigger function only be called by sensor driver while there is a new sample to fetch.

- Case 2: between `fetch accel` call and `fetch gyro` call
In this case, the sensor driver should ensure the `fetch accel` call should only check and clear `drdy_acc` bit, and `fetch gyro` call should also only check and clear `drdy_gyro` bit. With this guarantee, it can work normally like the case 1.

Reply:
> Dose `between fetch calls` mean between two `fetch accel` calls or between `fetch accel` call and `fetch gyro` call?
> 
> * Case 1: between two `fetch accel` calls
>   In this case, the sensing subsystem(polling mode) and data ready trigger function(data ready trigger mode) will make sure that there is a new sample between two `fetch accel` calls.
> 
> > * Polling mode:
> >   The freq of `fetch` called by Sensing subsystem is always less than sensor working ODR.
> > * Data ready trigger mode:
> >   The `fetch` is triggerd by the data ready trigger function which set by `sensor_trigger_set`.
> >   And data ready trigger function only be called by sensor driver while there is a new sample to fetch.
> 
> * Case 2: between `fetch accel` call and `fetch gyro` call
>   In this case, the sensor driver should ensure the `fetch accel` call should only check and clear `drdy_acc` bit, and `fetch gyro` call should also only check and clear `drdy_gyro` bit. With this guarantee, it can work normally like the case 1.

I'm talking about case 2, where you need to read accel and gyro. Most devices don't have separate data ready bits let alone registers. These status registers clear when read so when you fetch for the `accel` request you'll clear the status register and won't have data for the gyro. If you wait for new data for the gyro you'll either:
1. Get data from another frame (which may make the fusion of the 2 not very useful)
2. Have to wait a long time if the sensor requires warming up (some sensors take a long time to collect a sample for one-shot readings).

I didn't want to dive into it, but you can expand this further to the FIFO. Say you have an accel/gyro/die_temp driver that leverages a hardware FIFO. A single reading will contain samples from all 3 sensor types. If you have separate drivers like the current subsystem I'm not sure how you plan on fanning out the data. The accel phy_3d_sensor will get the full FIFO and will need to know about the other drivers. My model uses a single driver and 3 info nodes that are bound. This way the driver is aware of all the capabilities that are enabled and is able to service all 3 sensor types from one reading or the FIFO.

Reply:
For case 2, a lot of sensors (like BMI160, LSM6DSO) support separated data_rdy bits, so their existing sensor device driver can be used directly; for sensors which do not support separated data_rdy bits, the solution is simply add soft data_rdy bit in its sensor device driver implementation.
On FIFO, given today there is no existing Zephyr sensor device driver that supports FIFO, we did not plan to cover FIFO based combo sensor case in this PR. We can provide an example PR based on #60063 to show how phy_3d_sensor wrapper works with FIFO enabled sensor, formal PR will not cover FIFO until #60063 got merged.

Reply:
Please create and RFC for the proposal to support soft data ready bits per channel. I'm having a hard time seeing how this will scale and will be testable.

While I understand that you did not plan to cover the FIFO based sensor here, could you please explain how this will work. Currently there's a `struct device` created per sensor type. How will they be linked together?